contracts/pools/idle/StableSwapIDLE.vy

```diff
@@ -0,0 +1,998 @@
+    Ann: uint256 = A_ * N_COINS
+    c: uint256 = D
+    S_: uint256 = 0
+    _x: uint256 = 0
+    y_prev: uint256 = 0
+
+    for _i in range(N_COINS):
+        if _i == i:
+            _x = x
+        elif _i != j:
+            _x = xp_[_i]
+        else:
+            continue
+        S_ += _x
+        c = c * D / (_x * N_COINS)
+    c = c * D * A_PRECISION / (Ann * N_COINS)
+    b: uint256 = S_ + D * A_PRECISION / Ann  # - D
+    y: uint256 = D
+    for _i in range(255):
+        y_prev = y
+        y = (y*y + c) / (2 * y + b - D)
+        # Equality with the precision of 1
+        if y > y_prev:
+            if y - y_prev <= 1:
+                return y
+        else:
+            if y_prev - y <= 1:
+                return y
+    raise
+
+
+@view
+@external
+def get_dy(i: int128, j: int128, dx: uint256) -> uint256:
+    # dx and dy in c-units
+    rates: uint256[N_COINS] = self._stored_rates()
+    xp: uint256[N_COINS] = self._xp(rates)
+
+    x: uint256 = xp[i] + dx * rates[i] / PRECISION
+    y: uint256 = self.get_y(i, j, x, xp)
+    dy: uint256 = xp[j] - y
+    fee: uint256 = self.fee * dy / FEE_DENOMINATOR
+    return (dy - fee) * PRECISION / rates[j]
+
+
+@view
+@external
+def get_dx(i: int128, j: int128, dy: uint256) -> uint256:
+    # dx and dy in c-units
+    rates: uint256[N_COINS] = self._stored_rates()
+    xp: uint256[N_COINS] = self._xp(rates)
+
+    y: uint256 = xp[j] - (dy * FEE_DENOMINATOR / (FEE_DENOMINATOR - self.fee)) * rates[j] / PRECISION
+    x: uint256 = self.get_y(j, i, y, xp)
+    dx: uint256 = (x - xp[i]) * PRECISION / rates[i]
+    return dx
+
+
+@view
+@external
+def get_dy_underlying(i: int128, j: int128, dx: uint256) -> uint256:
+    # dx and dy in underlying units
+    rates: uint256[N_COINS] = self._stored_rates()
+    xp: uint256[N_COINS] = self._xp(rates)
+    precisions: uint256[N_COINS] = PRECISION_MUL
+
+    x: uint256 = xp[i] + dx * precisions[i]
+    y: uint256 = self.get_y(i, j, x, xp)
+    dy: uint256 = xp[j] - y - 1
+    _fee: uint256 = self.fee * dy / FEE_DENOMINATOR
+    return (dy - _fee) / precisions[j]
+
+
+@external
+@view
+def get_dx_underlying(i: int128, j: int128, dy: uint256) -> uint256:
+    # dx and dy in underlying units
+    rates: uint256[N_COINS] = self._stored_rates()
+    xp: uint256[N_COINS] = self._xp(rates)
+    precisions: uint256[N_COINS] = PRECISION_MUL
+
+    y: uint256 = xp[j] - (dy * FEE_DENOMINATOR / (FEE_DENOMINATOR - self.fee)) * precisions[j]
+    x: uint256 = self.get_y(j, i, y, xp)
+    dx: uint256 = (x - xp[i]) / precisions[i]
+    return dx
+
+
+@internal
+def _exchange(i: int128, j: int128, dx: uint256) -> uint256:
+    assert not self.is_killed
+    # dx and dy are in c-tokens
+
+    rates: uint256[N_COINS] = self._stored_rates()
+    xp: uint256[N_COINS] = self._xp(rates)
+
+    x: uint256 = xp[i] + dx * rates[i] / PRECISION
+    y: uint256 = self.get_y(i, j, x, xp)
+    dy: uint256 = xp[j] - y - 1  # -1 just in case there were some rounding errors
```

Comment:
There is some interesting stuff with rounding errors. Check how we do in base pool:
```python
    dy = (dy - dy_fee) * PRECISION / rates[j]
    assert dy >= min_dy, "Exchange resulted in fewer coins than expected"
    ...
    self.balances[j] = old_balances[j] - dy - dy_admin_fee
    ...
    transfer dy
```